src/openrct2/object/RideObject.cpp

```diff
@@ -523,305 +522,287 @@ uint8_t RideObject::CalculateNumHorizontalFrames(const rct_ride_entry_vehicle* v
+    car.sprite_height_positive = Json::GetNumber<uint8_t>(jCar["spriteHeightPositive"]);
+    car.animation = Json::GetNumber<uint8_t>(jCar["animation"]);
+    car.base_num_frames = Json::GetNumber<uint16_t>(jCar["baseNumFrames"]);
+    car.no_vehicle_images = Json::GetNumber<uint32_t>(jCar["numImages"]);
+    car.no_seating_rows = Json::GetNumber<uint8_t>(jCar["numSeatRows"]);
+    car.spinning_inertia = Json::GetNumber<uint8_t>(jCar["spinningInertia"]);
+    car.spinning_friction = Json::GetNumber<uint8_t>(jCar["spinningFriction"]);
+    car.friction_sound_id = Json::GetEnum<SoundId>(jCar["frictionSoundId"], SoundId::Null);
+    car.log_flume_reverser_vehicle_type = Json::GetNumber<uint8_t>(jCar["logFlumeReverserVehicleType"]);
+    car.sound_range = Json::GetNumber<uint8_t>(jCar["soundRange"], 255);
+    car.double_sound_frequency = Json::GetNumber<uint8_t>(jCar["doubleSoundFrequency"]);
+    car.powered_acceleration = Json::GetNumber<uint8_t>(jCar["poweredAcceleration"]);
+    car.powered_max_speed = Json::GetNumber<uint8_t>(jCar["poweredMaxSpeed"]);
+    car.car_visual = Json::GetNumber<uint8_t>(jCar["carVisual"]);
+    car.effect_visual = Json::GetNumber<uint8_t>(jCar["effectVisual"], 1);
+    car.draw_order = Json::GetNumber<uint8_t>(jCar["drawOrder"]);
+    car.num_vertical_frames_override = Json::GetNumber<uint8_t>(jCar["numVerticalFramesOverride"]);
+
+    auto jLoadingPositions = jCar["loadingPositions"];
+    if (jLoadingPositions.is_array())
     {
-        auto arr = ObjectJsonHelpers::GetJsonIntegerArray(jLoadingPositions);
-        for (auto x : arr)
+        for (auto& jPos : jLoadingPositions)
         {
-            peepLoadingPositions.push_back(x);
+            car.peep_loading_positions.push_back(Json::GetNumber<int8_t>(jPos));
         }
     }
     else
     {
-        auto& peepLoadingWaypoints = car.peep_loading_waypoints;
-        auto jLoadingWaypoints = json_object_get(jCar, "loadingWaypoints");
-        if (json_is_array(jLoadingWaypoints))
+        auto jLoadingWaypoints = jCar["loadingWaypoints"];
+        if (jLoadingWaypoints.is_array())
         {
             car.flags |= VEHICLE_ENTRY_FLAG_LOADING_WAYPOINTS;
+            car.peep_loading_waypoint_segments = Json::GetNumber<uint8_t>(jCar["numSegments"]);
 
-            auto numSegments = ObjectJsonHelpers::GetInteger(jCar, "numSegments");
-            car.peep_loading_waypoint_segments = numSegments;
-
-            size_t i;
-            json_t* route;
-            json_array_foreach(jLoadingWaypoints, i, route)
+            for (auto& jRoute : jLoadingWaypoints)
             {
-                if (json_is_array(route))
+                if (jRoute.is_array())
                 {
-                    size_t j;
-                    json_t* waypoint;
                     std::array<CoordsXY, 3> entry;
-                    json_array_foreach(route, j, waypoint)
+
+                    for (size_t j = 0; j < 3; j++)
```

Comment:
```suggestion
                    for (size_t j = 0; j < 3; ++j)
```
Why did we lose the range based here? Sounds dangerous to have the 3 hard coded

Reply:
`entry` is of fixed size 3 (see defintion on line 760), and access of points beyond that results in undefined behaviour. The existing code didn't check for this.